fabric-networking-v1/src/main/java/net/fabricmc/fabric/impl/networking/NetworkingInitializer.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.networking;
+
+import java.util.Collection;
+import java.util.Map;
+
+import com.google.common.collect.MapMaker;
+
+import net.minecraft.network.ClientConnection;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.PacketByteBuf;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.networking.v1.event.LoginQueryStartCallback;
+import net.fabricmc.fabric.api.networking.v1.receiver.ServerPacketReceiverRegistries;
+import net.fabricmc.fabric.api.networking.v1.sender.PacketByteBufs;
+import net.fabricmc.fabric.impl.networking.receiver.ServerPacketReceivers;
+
+public final class NetworkingInitializer implements ModInitializer {
+	private static NetworkingInitializer instance;
+	private final Map<ClientConnection, Identifier[]> connectionSpecificChannels = new MapMaker().weakKeys().initialCapacity(2).concurrencyLevel(2).makeMap();
```

Comment:
Have we considered the use of a mixin to attach this information to the objects instead of a weak map?

Reply:
this identifier array is relatively transient (present for a tiny fraction of a client connection's life time, and often unused if the client connection is just getting server info etc.), so I did not do mixins.